### PR TITLE
Setup Regal

### PR DESCRIPTION
**Pull request by Ptroger**
## Activate Linting in rego

This pr does three things:

- Configure and add linting 
- Add it to the CI
- Fix subsequent errors

### Configure and add linting
Regal has been added and configured in `.regal/config.yaml`. It's an extension of regal default configuration, where modification are explained in comments

OPA strict syntax check has been added in a makefile rule. It is important because Regal linter _expects_ this check passes.
!! This is **not** code format. This is some kind of a native linter that comes with OPA.

Regal will add this check to its pipeline in a future release, making the separate step unecessary.

### Add it to the CI
Two new rules:
- OPA check --strict 
- regal lint

### Fix subsequent errors
This one is quite wide. Rego was mostly rewritten. The big change is that we move from one god package to multiples. It introduces a few changes:
- There is a dependency management between packages
- Directory structure matches packages structures

Last but not least, the rego transpiler and wasm building had to be slightly changed at this scope:
Rego transpiler writes to 'main' package. This didn't change. But:
- criteria were moved to a separate package from 'main'
- So transpiled policy need to be importing 'criteria' package
- Also, './main.rego' was moved to './main/evaluate.rego'

We can be confident that core functionality was not broken in this rewriting by running tests over a launched stack.
`make armory-sdk/test/e2e`

### Notes
Please look at
- .[rego/config.yaml](https://github.com/narval-xyz/narval/blob/6f2f3062cec9199fff013f3ba7f5d3b96f216438/.regal/config.yaml#L1-L59) and read comments. That's were linter config choices are justified
- [the README file in `rego/`](https://github.com/narval-xyz/narval/blob/6f2f3062cec9199fff013f3ba7f5d3b96f216438/apps/policy-engine/src/resource/open-policy-agent/rego/README.md?plain=1#L1-L38)
- [policy-engine CI](https://github.com/narval-xyz/narval/blob/6f2f3062cec9199fff013f3ba7f5d3b96f216438/.github/workflows/policy-engine.yml#L114-L141)
- directory structure

### CODE REVIEW

File: apps/policy-engine/Makefile [hunk 1]
```diff
@@ -158,6 +159,22 @@ policy-engine/rego/format/check:
...
 
+policy-engine/rego/syntax/check:
+	opa check \
+		--strict \
+		${POLICY_ENGINE_PROJECT_DIR}/src/resource/open-policy-agent/rego
+
+policy-engine/rego/lint/fix:
+	regal fix \
+		${POLICY_ENGINE_PROJECT_DIR}/src/resource/open-policy-agent/rego \
+		-c .regal/config.yaml
```
Comment (wcalderipe) [inline apps/policy-engine/Makefile:170]: Could you please use the long version of the option name? It's a lot easier to wrap your head around script files when it's explicit what each option is.

I assume here it's `--config` but I'm not sure.

File: apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/checkIntentAmount_test.rego [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	checkIntentAmount({"currency": "fiat:usd", "operator": constants.operators.lessThan, "value": test_data.tenMaticValue}) with input as test_data.requestWithEip1559Transaction with data.entities as test_data.entities
+	checkIntentAmount({"currency": "fiat:usd", "operator": constants.operators.greaterThanOrEqual, "value": test_data.oneMaticValue}) with input as test_data.requestWithEip1559Transaction with data.entities as test_data.entities
+	checkIntentAmount({"currency": "fiat:usd", "operator": constants.operators.lessThanOrEqual, "value": test_data.oneMaticValue}) with input as test_data.requestWithEip1559Transaction with data.entities as test_data.entities
+}
+
+test_checkIntentAmount2 if {
+	requ := object.union(test_data.requestWithEip1559Transaction, {"intent": object.union(test_data.requestWithEip1559Transaction.intent, {"amount": "100000000000000000000000000000000000000000000000000000000000000000000"})})
+	checkIntentAmount({"operator": constants.operators.greaterThan, "value": "10"}) with input as requ with data.entities as test_data.entities
+}
```
Comment (wcalderipe) [inline apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/checkIntentAmount_test.rego:37]: Same feedback on these test (probably copied and pasted again) but they are only covering the happy path, and it's difficult to say what they're actually testing.

File: apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/checkRateLimit_test.rego [hunk 3]
```diff
@@ -35,7 +39,7 @@ rateFixedPeriodRequest = object.union(requestWithEip1559Transaction, {
 					"to": "eip155:137:0x000c0d191308a336356bee3813cc17f6868972c4",
 					"token": "eip155:137/erc20:0x2791bca1f2de4661ed88a30c99a7a9449aa84174",
 					"rates": {"fiat:usd": "0.99", "fiat:eur": "1.10"},
-					"timestamp": (getStartDateInNanoSeconds("1d") / 1000000) + ((60 * 60) * 1000), # current day plus 1 hour
+					"timestamp": (lib.getStartDateInNanoSeconds("1d") / 1000000) + ((60 * 60) * 1000), # current day plus 1 hour
```
Comment (wcalderipe) [inline apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/checkRateLimit_test.rego:42]: That's nice!

File: apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/policies/approvals_policy_test.rego [hunk 4]
```diff
@@ -1,8 +1,8 @@
-package main
+package armory.criteria
 
-import future.keywords.in
+import rego.v1
 
-permit[{"policyId": "approvalByUsers"}] = reason {
+permit[{"policyId": "approvalByUsers"}] := reason if {
```
Comment (wcalderipe) [inline apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/policies/approvals_policy_test.rego:5]: Why have we changed to dynamic assignment? Anything in special or just a Regal recommendation?

**Review by wcalderipe — COMMENTED** [review]
1/3 review parts. 

I'll continue it tomorrow in the morning.

**Review by wcalderipe — APPROVED** [review]
Good job. I appreciate your efforts to improve the quality of the Rego code, as it's a cornerstone in our stack.

I've reviewed all the files under the assumption that no critical logic has been altered. **If you've significantly refactored the logic of any Rego files, please leave a comment on those files, ping me, and I'll review them more thoroughly.**

File: apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/policies/approvals_policy_test.rego (hunk 4 above)
Comment (Ptroger) [reply apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/policies/approvals_policy_test.rego:5]: It will be mandatory soon in rego, and is already strongly recommended

File: apps/policy-engine/src/resource/open-policy-agent/rego/main/default_policies.rego [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+	}
+}
+
+forbid[{"policyId": "default-forbid-policy", "policyName": "Default Forbid Policy"}] := {
+	"type": "forbid",
+	"policyId": "default-forbid-policy",
+	"policyName": "Default Forbid Policy",
+	"approvalsSatisfied": [],
+	"approvalsMissing": [],
+}
```
[the remark this answers is not in the document]
Comment (Ptroger) [reply apps/policy-engine/src/resource/open-policy-agent/rego/main/default_policies.rego:24]: TLDR:
- It's not _really used_ anymore
- We still need at least one forbid rule so that rego can build without transpiled policies, and that's necessary for rego unit testing.

Remember. We are not using rego in a straight manner. We are transpiling the policies. What we wrote in our rego criteria are functions that will be used by policies. Yet, we need at least one forbid in our set for the main evaluation logic to correctly build.
I did try multiple ways to get rid of it, because it triggered me so much to have this, but couldn't find a way to do so, without breaking rego unit testing.

File: apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/userOperation/transfers/userOperation_transfer.rego [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+	validTransfers = [transfer |
+		some m in matches
+		transfer = m[0]
+		condition = m[1]
+		util.checkTransferAmount(transfer.amount, condition)
+	]
+
+	count(intentTransfers) == count(validTransfers)
+}
```
[the remark this answers is not in the document]
Comment (Ptroger) [reply apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/userOperation/transfers/userOperation_transfer.rego:46]: That's normal. I didn't wanted to spend time on userOperation criteria.

- It feels fishy and most likely doesn't need only a small refactor, but needs to be reshaped. Some functionality seems off, I don't understand why code was written twice, and some criteria will be killed as they don't make sense
- It's not really used as we didn't completely shaped 4337 support yet
- It's not a 1 day task, more like 1 week imo
- I've left it as is because I'm not sure what I should do with that right now

File: apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/userOperation/signMessage/message.rego [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	condition.value != constants.wildcard
+	condition.operator == constants.operators.equal
+	condition.value == intent.message
+}
+
+checkUserOperationMessage(intent, condition) if {
+	condition.value != constants.wildcard
+	condition.operator == constants.operators.has
+	contains(intent.message, condition.value)
+}
```
[the remark this answers is not in the document]
Comment (Ptroger) [reply apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/userOperation/signMessage/message.rego:25]: See comment below, I don't really want to touch user operation in rego right now

File: apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/checkIntentBeneficiary.rego [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package armory.criteria
+
+import rego.v1
+
+checkIntentBeneficiary(values) if {
```
[the remark this answers is not in the document]
Comment (Ptroger) [reply apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/checkIntentBeneficiary.rego:5]: No it's a 4337 domain terminology.
TLDR: A beneficiary is the person that will receive money because it paid on behalf of a user transaction fees for a transaction. That is the system that allows gasless transaction

File: apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/checkGasFeeAmount.rego [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	condition = getGasFeeAmountCondition(filters)
+	condition.operator == constants.operators.greaterThanOrEqual
+	to_number(condition.value) <= getGasFeeAmount(condition.currency)
+}
+
+checkGasFeeAmount(filters) if {
+	condition = getGasFeeAmountCondition(filters)
+	condition.operator == constants.operators.lessThanOrEqual
+	to_number(condition.value) >= getGasFeeAmount(condition.currency)
+}
```
[the remark this answers is not in the document]
Comment (Ptroger) [reply apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/checkGasFeeAmount.rego:88]: Yes agreed, although the method to getting to good test coverage IDK. I'd think we need to do it iteratively when we are in the good context. That's also a path to never getting it done. Wdyt ?

File: apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/checkDestinationAccountType.rego [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package armory.criteria
+
+import rego.v1
+
+import data.armory.entities
+
+checkDestinationAccountType(values) if {
+	destination = entities.buildIntentDestinationChainAccount(input.intent)
```
[the remark this answers is not in the document]
Comment (Ptroger) [reply apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/checkDestinationAccountType.rego:8]: Ty, I doubted a lot on correct namings for the libs, that's a nice pattern.

File: .regal/config.yaml [hunk 11]
```diff
@@ -0,0 +1,57 @@
+ignore:
+  # Ignore userOperation files as functionality is not certain yet.
+  files:
+    - '*userOperation*'
+rules:
+  style:
+    # This rule is disabled in tests to let us paste in big JSON payloads without having to worry about line length.
```
[the remark this answers is not in the document]
Comment (Ptroger) [reply .regal/config.yaml:7]: Ok so I may be agreeing here. But we virtually did it for every test. Meaning that if we enable this rule, there's half a day of work just to format big JSON payloads by splitting them

File: apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/checkAccount_test.rego [hunk 12]
```diff
@@ -0,0 +1,23 @@
+package armory.criteria
+
+import data.armory.entities
+import data.armory.test_data
+import rego.v1
+
+test_resource if {
```
[the remark this answers is not in the document]
Comment (Ptroger) [reply apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/checkAccount_test.rego:7]: Yes agreed, we need to keep track of that. Although reminder, this won't be done at the scope of this PR. Goal here is not to modify existing rego logic/testing, but to format it correctly.

File: apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/checkAssigned_test.rego [hunk 13]
```diff
@@ -0,0 +1,8 @@
+package armory.criteria
+
+import data.armory.test_data
+import rego.v1
+
+test_assigned_account if {
```
[the remark this answers is not in the document]
Comment (Ptroger) [reply apps/policy-engine/src/resource/open-policy-agent/rego/armory/criteria/__test__/checkAssigned_test.rego:6]: Yes we could, we can put a regex that checks that all test starts with 'test_' followed by a camelCase name

File: .regal/config.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (wcalderipe) [reply .regal/config.yaml:7]: So, let me step back and ask what's "big"?

Lengths are often to express horizontal size of a line. But maybe Regal uses it for vertical size instead.

File: .regal/config.yaml [hunk 14]
```diff
@@ -0,0 +1,69 @@
+ignore:
+  # Ignore userOperation files as functionality is not certain yet.
+  files:
+    - '*userOperation*'
```
Comment (mattschoch) [inline .regal/config.yaml:4]: Add to the ignore:
```
    - 'dist/**'
    - 'rego-build/**'
```

If you build locally, then I'm getting a bunch of lint errors from those; should be fine to ignore the dist stuff.

**Review by mattschoch — APPROVED** [review]
